### PR TITLE
[Router] Implement router backbone

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

The first step of https://github.com/sgl-project/sglang/issues/1732. See full context in the [design doc](https://docs.google.com/document/d/1cCqK3dh7ZR_rUPkcZT2cr0kLnAxv6_Sd-P1q37-3RNQ/edit?tab=t.0#bookmark=id.2244ui289zs1)
This PR implements:

1. Two-level HTTP architecture
2. Periodic Garbage Collection (GC)
3. Dynamic Scaling
4. Basic request routing (round robin and random)

## Modifications

1. Create `router/` folder under srt
2. Implement router and http server
3. Handle GC and dynamic scaling

## Usage

1. Launch workers

```python
export CUDA_VISIBLE_DEVICES=0; python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --host 127.0.0.1 --port 9000
export CUDA_VISIBLE_DEVICES=1; python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --host 127.0.0.1 --port 9002
```

2. Launch a router

> Note: for multi-node, replace `worker-urls` with accessible http endpoint

```python
python -m sglang.srt.router.launch_router --host 127.0.0.1 --port 8080 --policy round_robin --worker-urls http://127.0.0.1:9000 http://127.0.0.1:9002
```

3. Send a curl request to the router

```bash
curl -X POST http://127.0.0.1:8080/generate  -H "Content-Type: application/json" -d '{
    "text": "Once upon a time,",
    "sampling_params": {
      "max_new_tokens": 16,
      "temperature": 0
    }
  }'
```

## Tests

1. Add a ci test to test router with DP = 2, ensuring gsm8k score is reasonable.
2. Do a manual test to ensure the failure of a worker does not crash the router

## TODO

1. Implement cache-aware routing (collab w @Ying1123 @yichuan520030910320  )
3. Profile and improve performance
4. Add tests for fault tolerance (like killing a worker in the middle), but make sure the test is not flaky on CI
5. Add complete support of SGLang compatible apis (like `/health_generate` etc)
6. (stretched) Implement router in another language to reduce overhead (e.g. rust)

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: .github/workflows/pr-test-amd.yml
Comment (ByronHsu) [inline .github/workflows/pr-test-amd.yml:1]: will revert once review is done. this is for fast ci iteration

File: test/srt/test_router.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    def test_gsm8k(self):
+
+        parsed_url = urlparse(self.router_url)
+
+        host = parsed_url.hostname
+        port = parsed_url.port
+
+        data_path = "test.jsonl"
+
+        if is_local() is True:
```
Comment (ByronHsu) [inline test/srt/test_router.py:151]: ditto

File: .github/workflows/pr-test.yml [hunk 2]
```diff
@@ -242,32 +36,8 @@ jobs:
...
           pip install -e .
 
-      - name: Evaluate Accuracy (TP=2)
-        timeout-minutes: 20
-        run: |
-          cd test/srt
-          python3 test_moe_eval_accuracy_large.py
-      
-      - name: Evaluate MLA Accuracy (TP=2)
+      - name: Evaluate DP Router Accuracy (DP = 2)
```
Comment (ByronHsu) [inline .github/workflows/pr-test.yml:39]: ditto

File: .github/workflows/pr-test-amd.yml
Comment (merrymercy) [reply .github/workflows/pr-test-amd.yml:1]: this has been disabled

**Review by merrymercy — COMMENTED** [review]
Per our discussion offline, we would like to go with rust. Should we still merge this?

**Comment by ByronHsu** [comment]
Due to the low performance of 2-level HTTP with python server, i will try to re-implement in rust. Closing the issue for now!

**Pull request closed without merging**